### DISCUSSION TITLE
Thumbnail publisher increases CPU usage significantly

### DISCUSSION THREAD

**Discussion by hernanrz** [OvenMediaLabs/OvenMediaEngine#1447] [Help]
On our website, we have noticed that enabling the thumbnail publisher (see config) increases OME's cpu usage a lot, when compared to a script that periodically runs ffmpeg to generate thumbnails by pulling the hls stream from OME. Any reasons this might happen and how can we reduce it?

Server.xml https://gist.github.com/hernanrz/0defd82474647685bc02bfe7e50c513a

thumbnail script for comparison https://gist.github.com/hernanrz/f9db82df1796b95209b425d51cab372b

**Comment by bchah** [comment]
I recall @Keukhan may have answered this. The thumbnail publisher currently decodes every frame whether it is creating a thumbnail or not. Your approach is more efficient although I wonder if there's a cost to spawning many short-lived FFmpeg processes at once. You guys operate at scale so I guess it's OK?

Anyways, +1 for a more efficient thumbnail publisher. My application relies on "live" thumbnails and I definitely notice the impact on CPU.

**Reply by naanlizard** [reply]
We make screenshots and thumbnails every ten seconds with ffmpeg for up to a hundred streams and don't run into any real problems no. If we were worried about that we could just reduce the frequency of updates

**Reply by naanlizard** [reply]
For reference, our ffmpeg call that is run every 10s

`timeout 6s ffmpeg -y -i #{ENV["OME_HTTP_INTERNAL"] || 'http://localhost:799'}/live/#{stream}/llhls.m3u8 -frames:v 1 -update 1 -f mjpeg #{screenshot_path(stream)}.tmp -frames:v 1 -update 1 -s 558x302 -f webp #{thumbnail_path(stream, extension: "webp")}.tmp && mv #{screenshot_path(stream)}.tmp #{screenshot_path(stream)} && mv #{thumbnail_path(stream, extension: "webp")}.tmp #{thumbnail_path(stream, extension: "webp")}`

**Comment by hernanrz** [comment]
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/13426225/fe511363-bd44-4a62-ac85-6bee5ca9d4af)

Left hand side is OME with Thumbnail publisher disabled, blue line is Ruby Thumbnailer CPU usage, Right hand side is OME with Thumbnailer published enabeld, you can see it uses more CPU than the left side, it goes down as streams started to go offline

**Comment by getroot** [comment]
This is in our backlog. We will check to see if we can improve.

**Reply by bchah** [reply]
I have noticed that thumbnails, particularly HEVC, can max out a single thread and start to occupy more memory until all of the system memory is used:
![memory used](https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/51ce3037-2f10-44ba-a85e-8ab778eaeb9d)

See example log: 
[log.txt](https://github.com/AirenSoft/OvenMediaEngine/files/13445936/log.txt)

Besides the error messages in the log and the pegged CPU core, the outward symptom is that thumbnails appear to have more and more latency as the queue falls further behind. I wonder if it is possible to drop frames or 'flush' the queue to keep these crashes from happening?

**Reply by getroot** [reply]
HEVC software decoders can really use a lot of CPU. Flushing the queue is essentially the same as deleting the stream because it causes errors in the stream. Issues related to thumbnail performance will be looked into in depth by @Keukhan in the future when we have time.

**Reply by bchah** [reply]
Hello @getroot I am updating you here that this issue still occurs in the same manner after commit 9e9be09. It still seems to just use one thread. Is there anything in the OME config that needs to be set to use multiple threads on the Thumbnail publisher?

```
[2024-06-17 17:10:23.791] W [InboundWorker:6602] ManagedQueue | managed_queue.h:444  | [125] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_hevc_256 size has exceeded the threshold: queue: 1396, threshold: 500, peak: 1396
```

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/157ac6c5-2aee-456f-bd36-61cb15915ca3)

**Reply by bchah** [reply]
Hello, I am following up after the latest 0.16.8 release to confirm that the thumbnail publisher still crashes the system due to the ManagedQueue filling up. I know you don't consider this a bug but it is concerning because OME does not react to ManagedQueue exceeding the threshold, making a crash inevitable after a certain period of time because it will eventually eat all of the available RAM. Some more possible ideas to help this:

- For the thumbnail publisher, if the managed queue exceeds the threshold, could we pause the decoding to prevent system resources overrun and continue serving the last processed thumbnail image until there is room available in the queue for new frames?

- Recently @Keukhan expanded the thread count of the decoders to 2. Can this expand further to utilize any available CPU threads? I am still seeing the queue exceeding the threshold while available CPU resources are sitting idle.

- Could the Alert module be expanded to notify when an incoming stream overruns a transcoder queue, to allow us to take action? 

- It would also help if we could pause or stop/start a transcode separate from the actual stream itself. Perhaps related to the alert module, we could trigger a transcode to stop or pause using our own logic.

I understand this is not a trivial request! But I also believe that any condition which causes a memory leak like this needs to be addressed. I hope you can take another look in this area soon. Thank you.

**Reply by naanlizard** [reply]
Fwiw, we're still manually creating thumbnails with ffmpeg in production. I can post our settings if it helps, or benchmark our system against the built in one, basically anything other than write code for ome

Would absolutely love if ome handled screenshots for us

**Reply by bchah** [reply]
Thanks @naanlizard, I'm curious... how does that scale for you? Is there a way to keep the same instance going and pipe multiple streams in/out? We generally support 5-25 live streams at each customer site. I'd be weary of spinning up that many FFmpegs every 2 seconds for a 1-frame output. But if it works... :)

**Reply by naanlizard** [reply]
5-50 streams, typically. We update our screenshots every 10s, which is enough for our purposes, but I don't see why 2s would be unreasonable if you have the CPU for it

I'm not aware of a way to have ffmpeg connect and spit out a frame every X seconds, but it may exist, I'll google a bit 

We have a docker container that every 10s runs an ffmpeg command IIRC.

**Reply by bchah** [reply]
Looks like we discussed this last year in this very thread 😂

This is a cut down version of the command you're using
```
ffmpeg -y -i "hlslink.com/llhls.m3u8" -frames:v 1 -update 1 -s 640x360 -f mjpeg out.jpg
```
It works but the 16+ seconds latency on the output thumbnail makes me think FFmpeg is only reading traditional HLS. Any idea how to make it read LL-HLS?

**Reply by naanlizard** [reply]
Let me test but I'm pretty sure if you feed it an llhls link it'll be low ish latency

**Reply by bchah** [reply]
Appreciate it! I only tested the macOS homebrew version of FFmpeg 7.0 so far. Could be that this particular build is missing a piece too.

**Comment by naanlizard** [comment]
Hmm, at least my version of ffmpeg isn't doing any better than yours, even
after playing with a bunch of flags. I can document more later but gotta
run for now

**Comment by Keukhan** [comment]
@hernanrz  @bchah  @naanlizard 

Thank you for your patience. An option to decode only keyframes has been added to significantly reduce CPU usage when only the thumbnail publisher is used. It has been added to the master branch, but it has not been released yet. Please test it and let us know if you have any suggestions for improvements. 

Please refer to the link below for usage instructions.

Gitbook : https://airensoft.gitbook.io/ovenmediaengine/dev/transcoding#keyframe-decoding-only
Github : https://github.com/AirenSoft/OvenMediaEngine/commit/4dff75a63e239445313bdbc1b728d0db3c708dcd

Below are the results from my testing

**Before**
![image](https://github.com/user-attachments/assets/d933633a-4cc8-47a5-b911-9f0b984d8f3b)

**After**
![image](https://github.com/user-attachments/assets/749cc28d-1084-4e4b-bf52-9d4dead4846f)




@bchah 

The feature to configure the number of decoding threads you requested has also been added

Gitbook : https://airensoft.gitbook.io/ovenmediaengine/dev/transcoding#decoding-thread-configuration
Github : https://github.com/AirenSoft/OvenMediaEngine/commit/c0a8ead2fb8053dfa998565a4bdf95eccc366251

**Reply by naanlizard** [reply]
Some initial thoughts: 

With KFI=2 for my input stream and the following config, I'm getting about 16s latency between live and the screenshot. Does that sound correct? Is there any way to improve that latency? It's not a big deal, but worth asking.

Is there a performant way to serve two sizes of the same screenshot with OME? Or should I just use something like [imgproxy](https://imgproxy.net/)?

Is there a way to save these screenshots to the filesystem with OME directly? Or again would I need a third party tool?

```xml
				<Application>
					<Name>live</Name>
					<!-- Application type (live/vod) -->
					<Type>live</Type>
					<OutputProfiles>

					    <!-- 
						Common setting for decoders. Decodes is optional.
						-->

						<Decodes>
							<KeyframeOnlyIfNeed>true</KeyframeOnlyIfNeed>
						</Decodes>


						<!-- Enable this configuration if you want to hardware acceleration using GPU -->
						<HardwareAcceleration>false</HardwareAcceleration>

						<OutputProfile>
							<Name>bypass_stream</Name>
							<OutputStreamName>${OriginStreamName}</OutputStreamName>
							<Encodes>
								<Audio>
									<Bypass>true</Bypass>
								</Audio>
								<Video>
									<Bypass>true</Bypass>
								</Video>
								<Audio>
									<Codec>opus</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
								</Audio>
								<Image>
									<Codec>jpeg</Codec>
									<Framerate>0.1</Framerate>
								</Image>
							</Encodes>
						</OutputProfile>
```

**Comment by naanlizard** [comment]
Wow, perfect timing! We'll try it asap

I was just fighting with our current thumbnail setup

**Comment by naanlizard** [comment]
A question about the implementation:

When the option is enabled and ome only decodes key frames, how does the
screenshot output profile framerate setting interact?

If my stream has a keyframe interval of 1s and the target output is 1s, I
assume it works as expected and the image updates every 1s

If the KFI is 10s and the framerate is set to 1s, I think it would update
every 10s?

Similarly, KFI=1s, framerate 10s, image updates every 10s?

I think the flag name is confusing and will think of a better one.
KeyframeOnlyImages? Something like that

I will try a pull request with maybe a better name and also update the
gitbook

**Reply by Keukhan** [reply]
In my stream, when the Keyframe Interval (KFI) is set to 10 and the image framerate is 1, the image updates only once every 10 seconds, which makes it hard to get accurate keyframes.

When KFI is set to 1 and the framerate is 10, decoding happens 10 times, but the image still only updates every 10 seconds. It would be possible to improve this by skipping 9 of the decodes, but that hasn't been developed yet.

If you have any good name ideas, please share them with us. There have been many comments internally that the term isn’t  intuitive. :)

**Reply by bchah** [reply]
Thank you Keukhan! How about _PreferKeyframeOnlyDecoding_? Or, since it is already part of the `<Decoding>` block, simply _PreferKeyframes_?

**Reply by naanlizard** [reply]
I think "Prefer" is too ambiguous - as I understand the feature, it ONLY decodes keyframes, so there's no preference involved (as opposed to say, preferring vanilla, but if vanilla isn't available, you'll eat chocolate)

`DecodeOnlyKeyframes`? `OnlyKeyframes`?

**Reply by Keukhan** [reply]
You all know my english isn't very good. so, i don't know feeling of words. If you all pick one name and let me know, I’ll change the option to that. Even if you pick something like < BchahAndNaanlizard >, I’ll use it.

Currently, < Decodes > . < KeyframeOnlyIfNeed > is set, meaning it will only decode keyframes when it's a good situation to do so. If the video encoding is in OutputProfile, this option will automatically turn off.

**Reply by bchah** [reply]
Thanks Keukhan. You're the best. All I really care is that you keep drinking water! 😀

**Reply by Keukhan** [reply]
> Thanks Keukhan. You're the best. All I really care is that you keep drinking water! 😀

Last weekend, my wife was wearing a Streamwell T-shirt... It’s a rare item, so I had hiding it... but I failed.

**Reply by naanlizard** [reply]
Amazing work, performance is fantastic in our testing.

One question: Does OnlyKeyframes always generate a thumbnail from the first keyframe? For example, you have an fps for your thumbnails set to .1 (10s between thumbs), and the keyframes appear at 0s, 4s, 8s, 12s, etc (I'm not sure how keyframes work honestly, but I assume the stream starts with one). When does the first thumbnail get generated?

Some suggestions, I can open new tickets or discussions if you like

1. Returning a 404 instead of text makes more sense to me, and I think is the standard
2. Caching the last image generated for a stream would be convenient - if a stream is offline, return the last thumbnail from when that stream WAS live (assuming it ever was), or if a thumb hasn't been generated yet because the stream just went live, return the old thumb. 
3. One way to support 2, and serve files directly, would be to save thumbs to the filesystem when generated. Can OME do this? This would let us serve directly via nginx instead of load on OME, and obviate #2 (there's a fun vocab word, in this context, it means to make #2 unnecessary)
4. Is it possible to set multiple outputs for images? e.g. one output at full resolution, one at thumbnail size?

**Reply by bchah** [reply]
Seconding the lizard. My Dech264 processes went from ~12.5% CPU per stream to ~0.5%.

Regarding #2,3... There are many good reasons to not give users direct access to any OME endpoint including the thumbnail publisher. In my application our API does something similar to what you suggest:

- Receives an initial request for a thumbnail
- Requests the thumbnail from the OME API
- If a thumbnail was retrieved, caches it to disk, otherwise serves a placeholder 'offline' thumbnail (in your case you could just keep serving the last cached thumbnail if there is one)
- If another request for that same thumbnail comes back within the user-specified thumbnail refresh interval, serves the thumbnail from disk rather than hitting OME a second time, otherwise we begin again

Regarding #4... I know you want multiple outputs per decode which is a good feature request, but now given the significant decrease in required resources for decoding, you could try running two Image encodes. One called ${OriginStreamName}_full, one called ${OriginStreamName}_thumb.

My biggest feature request around this would be to let us pipe the outputs of any encoder to FFmpeg. Then we can apply our own mad science 🧪

**Comment by naanlizard** [comment]
Lastly, and I'll probably open a new ticket, but are PNG and jpeg the only
options for image output now? Any hope of webp?

**Reply by Keukhan** [reply]
Please create a new ticket for the WEBP-related request. I will include it in the next plan.

**Reply by naanlizard** [reply]
I ask just for my own planning, not to pressure you - is there an ETA, even very rough for WEBP support? e.g. 1 week, 1 month, 1 year?

**Reply by Keukhan** [reply]
The development is expected to take less than a week.  But we need to talk internally to decide when we can start working on WEBP codec support.

**Reply by bchah** [reply]
Rather than tackling one format request at a time, allowing us to pipe encoder outputs to FFmpeg or ImageMagick would give us a ton of flexibility here. I imagine a world where encoder configs in OME have an option like this:

```
 <Encodes>
   <Video>
    <Name>original_video</Name>
    <Bypass>true</Bypass>
    <FFmpeg>ffmpeg -f format -i - output_file</FFmpeg>
  </Video>
</Encodes>
```

**Comment by bchah** [comment]
I'm excited to update our shipping version of OME - hopefully there is an official release containing these and the other recent improvements soon! 👍
